### PR TITLE
Add loop event to engine

### PR DESCRIPTION
**Pull request by cassava**
This is required by the UI for properly changing simulation speed.

### CODE REVIEW

**Review by clonker — APPROVED** [review]

File: engine/src/simulation_context.hpp [hunk 1]
```diff
@@ -355,6 +356,7 @@ struct SimulationContext {
   bool pause_execution{false};
 
   // Events
+  std::shared_ptr<events::LoopCallback> callback_loop;
```
[the remark this answers is not in the document]
Comment (clonker) [reply engine/src/simulation_context.hpp:359]: personal feeling: shared-ptr is often overused and there's nothing inherently bad regarding the use of raw pointers if there's a clear ownership and therefore well-defined construction and destruction sequence

File: engine/src/simulation.cpp [hunk 2]
```diff
@@ -387,6 +387,7 @@ StateId SimulationMachine::Connect::impl(SimulationContext& ctx) {
     ctx.coordinator->enroll(r);
 
     // Events:
+    ctx.callback_loop = r.register_event<events::LoopFactory>();
```
[the remark this answers is not in the document]
Comment (clonker) [reply engine/src/simulation.cpp:390]: :+1:  i tried to indicate that in my comment, it was genuine curiosity if we had _actual_ co-ownership that makes shared_ptr a necessity in this case

File: engine/src/simulation_context.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (cassava) [reply engine/src/simulation_context.hpp:359]: Agreed.

File: engine/src/simulation.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (cassava) [reply engine/src/simulation.cpp:390]: I created a discussion for addressing this question.

**Pull request merged**